### PR TITLE
Refs #20010 -- Unified DatabaseOperations.last_executed_query() on Oracle with other db backends.

### PR DESCRIPTION
**Pull request by felixxm**

### CODE REVIEW

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
I added tests that cover both cases in `DatabaseOperations.last_executed_query()` on Oracle.

File: tests/backends/tests.py [hunk 1]
```diff
@@ -73,6 +73,37 @@ def test_query_encoding(self):
...
+            sql, params = qs.query.sql_with_params()
+            cursor = qs.query.get_compiler(DEFAULT_DB_ALIAS).execute_sql(CURSOR)
+            self.assertEqual(
+                cursor.db.ops.last_executed_query(cursor, sql, params),
+                str(qs.query),
+            )
+
+    @skipUnlessDBFeature('supports_paramstyle_pyformat')
+    def test_last_executed_query_dict(self):
+        # last_executed_query() handle dicts.
```
Comment (felixxm) [inline tests/backends/tests.py]: I will remove this useless comment before merge.

**Comment by felixxm** [comment]
@charettes Can you take a look again?

File: django/db/backends/oracle/operations.py [hunk 2]
```diff
@@ -258,9 +258,16 @@ def last_executed_query(self, cursor, sql, params):
...
-        # Unlike Psycopg's `query` and MySQLdb`'s `_executed`, CxOracle's
-        # `statement` doesn't contain the query parameters. refs #20010.
-        return super().last_executed_query(cursor, statement, params)
+        # Unlike Psycopg's `query` and MySQLdb`'s `_executed`, cx-Oracle's
+        # `statement` doesn't contain the query parameters. Substitute
+        # parameters manually.
+        if isinstance(params, (tuple, list)):
+            for i, param in enumerate(params):
+                statement = statement.replace(':arg%d' % i, force_str(param, errors='replace'))
+        elif hasattr(params, 'keys'):
```
Comment (charettes) [inline django/db/backends/oracle/operations.py]: It feels a bit odd to _duck-type_ on `.keys` but use `.items` in the loop; maybe `hasattr(params, 'items')` should be used instead? Is there a reason not to use `isinstance(params, dict)` instead?

Comment (felixxm) [reply django/db/backends/oracle/operations.py]: :thinking: Probably not, I change this :+1:

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Pull request merged**